auth/cache.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// Write saves HCP auth data in a common location in the home directory.
+func Write(token oauth2.Token) error {
+	credentialPath, credentialDirectory, err := getCredentialPaths()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve credential path and directory: %v", err)
+	}
+
+	err = os.MkdirAll(credentialDirectory, directoryPermissions)
+	if err != nil {
+		return fmt.Errorf("failed to create credential directory: %v", err)
+	}
+
+	// Write access token, refresh_token, expiry, max age to credentials file.
+	credentials := &Cache{
+		AccessToken:  token.AccessToken,
+		RefreshToken: token.RefreshToken,
+		Expiry:       token.Expiry,
+		MaxAge:       MaxAge,
+	}
+
+	credentialsJSON, err := json.Marshal(credentials)
+	if err != nil {
+		return fmt.Errorf("failed to marshal the struct to json: %v", err)
+	}
+
+	err = os.WriteFile(credentialPath, credentialsJSON, directoryPermissions)
+	if err != nil {
+		return fmt.Errorf("failed to write credentials to the cache file: %v", err)
+	}
+
+	return nil
+}
+
+// Read opens the saved HCP auth data and returns the token.
+func Read() (*Cache, error) {
+	credentialPath, _, err := getCredentialPaths()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve credential path and directory: %v", err)
+	}
+
+	rawJSON, err := os.ReadFile(credentialPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file from user's credential path: %v", err)
+	}
+
+	cacheFromJSON, err := jsonToToken(rawJSON)
+	if err != nil {
+		return nil, fmt.Errorf("bad format: %v", err)
+	}
+
+	return cacheFromJSON, nil
+}
+
+// getDirectory returns the complete credential path and directory.
+func getCredentialPaths() (credentialPath string, credentialDirectory string, err error) {
+	// Get the user's home directory.
+	userHome, err := os.UserHomeDir()
+	if err != nil {
+		return "", "", fmt.Errorf("failed to retrieve user's home directory path: %v", err)
+	}
+
+	directoryName := defaultDirectory
+	// If in test mode, use test directory.
+	if testMode, ok := os.LookupEnv(envVarCacheTestMode); ok {
+		if testMode == "true" {
+			directoryName = testDirectory
+		}
+	}
+
+	// Determine absolute path to config file and directory.
+	credentialDirectory = filepath.Join(userHome, directoryName)
+	credentialPath = filepath.Join(userHome, directoryName, fileName)
+
+	return credentialPath, credentialDirectory, nil
+}
+
+// jsonToToken converts the raw JSON into the Cache struct.
+func jsonToToken(rawData []byte) (*Cache, error) {
+
+	var cacheFromJSON Cache
+	err := json.Unmarshal(rawData, &cacheFromJSON)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal the raw data to json: %v", err)
+	}
+
+	if cacheFromJSON.AccessToken == "" {
+		return nil, errors.New("failed to get cache access token")
+	}
+
+	if cacheFromJSON.RefreshToken == "" {
+		return nil, errors.New("failed to get cache refresh token")
+	}
+
+	return &cacheFromJSON, nil
+}
+
+// tokenToJson converts the OAuth2 token to raw JSON.
```

Comment:
🤔 I wonder if we should make these conversions one-to-one to be more explicit about how the data is being transformed. 
- tokenToCache
- CacheToJSON
- jsonToCache
- CacheToToken

```suggestion
// tokenToJson converts the OAuth2 token first to a Cache and then to raw JSON.
```

Reply:
I think I'm a little confused with this one 🤔 Are you suggesting to have four functions? Wouldn't some of them just solely run the `Marshal` or `Unmarshal` functions?

Reply:
Yeah, probably more pedantic than necessary. I think what I'm getting at is maybe we just need a comment explaining the difference between the Cache and the Token, when you might need one vs. the other.